sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CertificateValidationCheck.cs

```diff
@@ -0,0 +1,310 @@
+            {
+                c.VisitedMethods.Add(containingMethodDeclaration);
+                var containingMethod = c.Context.SemanticModel.GetDeclaredSymbol(containingMethodDeclaration);
+                var paramSymbol = containingMethod.Parameters.Single(x => x.Name == param.Identifier.ValueText);
+                foreach (var invocation in FindInvocationList(c.Context, FindRootClass(param), containingMethod))
+                {
+                    var methodParamLookup = new CSharpMethodParameterLookup(invocation.ArgumentList, containingMethod);
+                    ArgumentSyntax argument;
+                    if (methodParamLookup.TryGetSymbolParameter(paramSymbol, out argument))
+                    {
+                        ret.AddRange(CallStackSublocations(c, argument.Expression));
+                    }
+                }
+            }
+            return ret.ToImmutable();
+        }
+
+        private static ImmutableArray<Location> VariableLocations(InspectionContext c, VariableDeclaratorSyntax variable)
+        {
+            var allAssignedExpressions = new List<ExpressionSyntax>();
+            var parentBlock = variable.FirstAncestorOrSelf<BlockSyntax>();
+            if (parentBlock != null)
+            {
+                allAssignedExpressions.AddRange(parentBlock.DescendantNodes().OfType<AssignmentExpressionSyntax>()
+                    .Where(x => x.Left is IdentifierNameSyntax Ident && Ident.Identifier.ValueText == variable.Identifier.ValueText)
+                    .Select(x => x.Right));
+            }
+            if (variable.Initializer != null)       //Declarator initializer is counted as (default) assignment as well
+            {
+                allAssignedExpressions.Add(variable.Initializer.Value);
+            }
+            return MultiExpressionSublocations(c, allAssignedExpressions);
+        }
+
+        private static ImmutableArray<Location> InvocationLocations(InspectionContext c, MethodDeclarationSyntax method)
+        {
+            var returnExpressionSublocationsList = method.Body.DescendantNodes()
+                .OfType<ReturnStatementSyntax>()
+                .Select(x => x.Expression)
+                .Where(x => !IsVisited(c, x));      //Ignore all return statements with recursive call. Result depends on returns that could return compilant validator.
+            return MultiExpressionSublocations(c, returnExpressionSublocationsList);
+        }
+
+        private static bool IsVisited(InspectionContext c, ExpressionSyntax expression)
+        {
+            if (expression is InvocationExpressionSyntax invocation)
+            {
+                var MS = c.Context.SemanticModel.GetSymbolInfo(invocation).Symbol;
+                return MS != null && MS.DeclaringSyntaxReferences.Length != 0 && MS.DeclaringSyntaxReferences.Select(x => x.GetSyntax() as MethodDeclarationSyntax).Any(x => x != null && c.VisitedMethods.Contains(x));
```

Comment:
_minor_: we tend to use `!xxx.IsEmpty` syntax for readibility instead of `xxx.Length != 0`